### PR TITLE
[SDK-17] Optimize sdk tests via reducing fixture times

### PR DESCRIPTION
**Pull request by vbrodsky**
Story: https://labelbox.atlassian.net/browse/SDK-17

The goal of this story was to reduce time spent in test setup (i.e. fixtures)

BEFORE estimate of average setup time over 5 slowest features on stage: 300s
AFTER: 224s

BEFORE: total running time of a successful run 13.5 mins (12 reruns)
AFTER: 11.5 mins (15 reruns)

Updates and improvements:
- Replaced `configured_project`(14 data rows) with `configured_project_with_one_data_row` where possible
  - According to my observation, each data row creation takes 1s and we create about 6K rows via the configured_project fixture during a typical run with RERUNs 
  - The less data rows we create, the less chance of data row creation timing out
- Reduced wait for data row creation from default 30s to 3s
  - I have observed that we usually create data rows, then proceed to create batch. Inside create batch, we wait for data rows completion with a sleep time of 30s. We usually never have a completion right away, so this adds 30s run time. By reducing the sleep time to 3s I am able to reduce running time of data row creation from ~ 30s to 15-20s per batch
- Added a way to record slowest-running individual fixtures (codefresh provides slowest setup times per test only, not cumulative over fixtures)
  - Optional feature, off by default
  - Currently per worker, need to be post-processed manually to get global total  
- Provided a way to customize data row ids logic for the `prediction_id_mapping` fixtures, drastically speed up tests not requiring actual data row creation
  - We should be able to extend this in another PR to reduce number of datarows created in configured_project 
- Reliability improvements
  - Removed some more data leaks
  - Another attempt to fix flaky `test_filtering`
  - Fixed possibly flaky test `test_user_and_org.`

IMPROVEMENT STORIES
- https://labelbox.atlassian.net/browse/SDK-203 - refactor project fixtures
- https://labelbox.atlassian.net/browse/SDK-204 - reduce number of data rows created in configured_project for prediction tests

### CODE REVIEW

File: tests/integration/export_v2/conftest.py [hunk 1]
```diff
@@ -297,7 +297,7 @@ def configured_project_with_ontology(client, initial_dataset, ontology,
 
 
 @pytest.fixture
-def configured_project_without_data_rows(client, ontology, rand_gen):
+def configured_project_with_one_data_row(client, ontology, rand_gen):
```
Comment (attila-papai) [inline tests/integration/export_v2/conftest.py:300]: this does not create a data row, should we revert the rename?

File: tests/integration/test_filtering.py [hunk 2]
```diff
@@ -24,18 +24,15 @@ def project_to_test_where(client, rand_gen):
 
 # Avoid assertions using equality to prevent intermittent failures due to
 # other builds simultaneously adding projects to test org
-def test_where(client, image_url, project_to_test_where, rand_gen):
+def test_where(client, project_to_test_where):
     p_a, p_b, p_c = project_to_test_where
-    p_a_name, p_b_name, p_c_name = [p.name for p in [p_a, p_b, p_c]]
+    p_a_name, p_b_name, _ = [p.name for p in [p_a, p_b, p_c]]
```
Comment (attila-papai) [inline tests/integration/test_filtering.py:29]: you can remove `_` if you remove `p_c` from the array

File: tests/integration/annotation_import/conftest.py [hunk 3]
```diff
@@ -556,27 +571,72 @@ def dataset_conversation_entity(client, rand_gen, conversation_entity_data_row,
...
+Please note that this fixture now offers the flexibility to configure three different strategies for generating data row ids for predictions:
+Default(configured_project fixture):
+    configured_project that generates a data row for each member of ontology.
+    This makes sure each prediction has its own data row id. This is applicable to prediction upload cases when last label overwrites existing ones
+
+Optimized Strategy (configured_project_with_one_data_row fixture):
+    This fixture has only one data row and all predictions will be mapped to it
+
+Custom Data Row IDs Strategy:
+    Individuals can create their own fixture to supply data row ids.
```
Comment (kkim-labelbox) [inline tests/integration/annotation_import/conftest.py:617]: Let's specify that this is only used for tests that require "data row ids", and not actual data rows created

File: tests/integration/conftest.py [hunk 4]
```diff
@@ -807,3 +818,33 @@ def upload_invalid_data_rows_for_dataset(dataset: Dataset):
...
+
+
+@pytest.hookimpl(hookwrapper=True)
+def pytest_fixture_setup(fixturedef):
+    start = time.time()
+    yield
+    end = time.time()
+
+    exec_time = end - start
+    if "FIXTURE_PROFILE" in os.environ:
```
Comment (kkim-labelbox) [inline tests/integration/conftest.py:834]: Let's make sure to include this in codefresh & github

File: tests/integration/annotation_import/conftest.py [hunk 5]
```diff
@@ -478,33 +479,47 @@ def func(project):
...
     project.setup(editor, ontology)
+
     data_row_ids = []
 
     for _ in range(len(ontology['tools']) + len(ontology['classifications'])):
         data_row_ids.append(dataset.create_data_row(row_data=image_url).uid)
+    project._wait_until_data_rows_are_processed(data_row_ids=data_row_ids,
+                                                sleep_interval=3)
+
     project.create_batch(
```
[the remark this answers is not in the document]
Comment (vbrodsky) [reply tests/integration/annotation_import/conftest.py:516]: yes

File: tests/integration/conftest.py [hunk 6]
```diff
@@ -386,11 +388,25 @@ def initial_dataset(client, rand_gen):
     dataset = client.create_dataset(name=rand_gen(str))
     yield dataset
 
+    dataset.delete()
+
+
+@pytest.fixture
+def project_with_ontology(project):
```
[the remark this answers is not in the document]
Comment (vbrodsky) [reply tests/integration/conftest.py:395]: k

File: tests/integration/conftest.py [hunk 7]
```diff
@@ -399,14 +415,7 @@ def configured_project(project, initial_dataset, client, rand_gen, image_url):
...
     project.data_row_ids = [data_row_id]
 
-    editor = list(
-        project.client.get_labeling_frontends(
-            where=LabelingFrontend.name == "editor"))[0]
-    empty_ontology = {"tools": [], "classifications": []}
-    project.setup(editor, empty_ontology)
     yield project
-    dataset.delete()
-    project.delete()
```
[the remark this answers is not in the document]
Comment (vbrodsky) [reply tests/integration/conftest.py:409]: I will add `batch.delete()` the rest of resources now come from other fixtures

File: tests/integration/conftest.py [hunk 8]
```diff
@@ -417,16 +426,21 @@ def configured_project_with_label(client, rand_gen, image_url, project, dataset,
...
         [data_row.uid],  # sample of data row objects
         5  # priority between 1(Highest) - 5(lowest)
     )
+    print("create_batch took: ", time.time() - start_time)
     ontology = _setup_ontology(project)
+    print("setup ontology took: ", time.time() - start_time)
     label = _create_label(project, data_row, ontology,
                           wait_for_label_processing)
-
+    print("create_label took: ", time.time() - start_time)
```
[the remark this answers is not in the document]
Comment (vbrodsky) [reply tests/integration/conftest.py:443]: no will remove

File: tests/integration/conftest.py [hunk 9]
```diff
@@ -442,10 +456,8 @@ def configured_batch_project_with_label(project, dataset, data_row,
     One label is already created and yielded when using fixture
     """
     data_rows = [dr.uid for dr in list(dataset.data_rows())]
-    project._wait_until_data_rows_are_processed(
-        data_row_ids=data_rows,
-        wait_processing_max_seconds=DATA_ROW_PROCESSING_WAIT_TIMEOUT_SECONDS,
```
[the remark this answers is not in the document]
Comment (vbrodsky) [reply tests/integration/conftest.py:447]: k

File: tests/integration/export_v2/conftest.py (hunk 1 above)
Comment (vbrodsky) [reply tests/integration/export_v2/conftest.py:300]: another good find

**Comment by vbrodsky** [comment]
@Attila @Kevin implemented all changes on this PR except from one batch deletion as it was giving me an error `labelbox.exceptions.OperationNotAllowedException: You can't delete batches that have labels.` I think not worth chasing in this case. Also did not follow up on adding the FIXTURE_PROFILE key as discussed with Kevin
Ready for your review

**Review by kkim-labelbox — APPROVED** [review]
LGTM! Nice work. Huge improvements!

**Pull request merged**